### ISSUE TITLE
Download stuck in "Download in progress"

### ISSUE THREAD

**Issue by MasothU** [mediathekview/zapp#202]
Hello,
I started a few downloads from the Mediathek. Some of them are not downloaded fully and I can't stop the download in zapp. It says "Download läuft..." and when I click on stop, nothing happens.
I already cleared the cache of zapp, Downloads and Download Manager, rebooted the phone. 
Any ideas?

Settings and version information:
Download to SD card
Version 3.5.0-beta.1 (41)
Samsung Galaxy S8, Android 9

**Comment by cemrich** [comment]
I tweaked the download behavior a bit and released v3.5.0 which should hit F-Droid soon. Hopefully this will resolve your issue. If you want a quick and dirty fix you can uninstall and reinstall Zapp (all your settings will be deleted this way).

**Comment by MasothU** [comment]
Thank you for your fast reply. Do you know when the new release will be available on F-Droid (one day, one week)? I'll test the new version and report here.
I guess it's a good idea to remove everything before and reinstall a clean version of zapp?

**Comment by cemrich** [comment]
The version has already been recognized by F-Droid but was not yet built. I think in a few days the new version should show up.


> I guess it's a good idea to remove everything before and reinstall a clean version of zapp?

No, as long as you don't switch between F-Droid and Github releases upgrades should run smoothly.

**Comment by cemrich** [comment]
Version 3.5.0 is available now. Does this fix your issue?

**Comment by justf0rfun** [comment]
I have the same issue and it is not resolved, yet. 

First I tried to download a video but my android device, but there was not enough free storage left. So the download was stuck in the middle. The download progress bar war about half filled. So I think there is half of the video file lying around somewhere in the filesystem. 

Unfortunately I was not able to find it. I also can't find completely downloaded videos anymore. Did the download target location change? For me it is important to have access to the videos because this let's me use them with other programs or move them somewhere else. Also it could help resolve issues manually. As video files can occupy significant amounts of storage and it can easily be forgotten by the user that there are (partially) downloaded videos which can only be seen in the app I think this is a significant problem. So the user must remember which videos he downloaded (partially) and find them in the mediathek again to delete them. So the user has to remember the right search terms and on series of shows the right episode (which moves down whenever a new episode is available). 

After that I tried a couple of things I don't remember exactly and now there is the download button with the text "Download läuft" a square on the left (probably the symbol for the stopping function) and an animated bar on the lower edge of the button.

Probably you had some work with the download function. Thank you for all the work you all do. 

Still maybe the download function should revised again. Maybe it is not a good solution to one button for downloading, (resuming?) and deleting a video. And I think it would be very useful to get filesystem access to the videos again.

Please tell me if you think that parts of this should become a new issue entry on its own (and which).

**Comment by cemrich** [comment]
> Unfortunately I was not able to find it

Thank you for your feedback!

You should be able to find downloaded videos very easily with any file explorer when filtering for "Videos". Unfortunately there is no clean way to open directories or download locations in Android.

I had to change download locations to make Zapp compatible with future versions of Android and following the Android developer guidelines. I realize that this change can be confusing.

**Comment by MasothU** [comment]
Hello cemrich,
sorry for the late reply. I was on vacation with only random access to slow internet and therefore couldn't check or download videos via zapp.
I did now and apart from some failed downloads (which I could restart and got successfully downloaded after a few tries) the error that a download was stuck in "Download in progress" did not occur to me.
I will do a batch download next week again, when new episodes of my shows have been aired and see if it still doesn't occur.

Another question. Is it possible to change the download button back to the old behaviour where you could directly select the download quality?
Doing batch downloads every week I now have to always click two times on every show. Only a small request :)

Regarding the question, where to find the videos on the external storage, in my case the path is the following (should be similar on other devices)
/storage/<name_of_external_sd_card_e.g._4707-08115>/Android/media/de.christinecoenen.code.zapp

A good file explorer for Android I can recommend is FX File Explorer

**Comment by justf0rfun** [comment]
I thought maybe there should be a "Downloads" view inside zapp. This would allow easy monitoring of all downloads. On the other side it could be questioned why zapp downloads itself instead of a download manager. The download manager which was default on my android is pretty bad. I don't know if this is the default on android in general. Actually what I like about downloading directly inside zapp is how easy and quickly accessible it is.

In my experience, or in my situation many download errors are caused by insufficient storage space. I think this could be handled better if the storage space for a download would be checked before starting the download and if the needed space can be reserved at begin of the download instead of downloading big parts of the file and then android going mad because there is no storage left.

Please tell if, if some of this should go into a separate issue.

Please take all of this as **constructive** critics.

**Comment by MasothU** [comment]
> 
> 
> I thought maybe there should be a "Downloads" view inside zapp. 

+1 from my side.
> On the other side it could be questioned why zapp downloads itself instead of a download manager. 

I guess it is just more easy to use the already available download manager instead of rebuilding it inside zapp


> In my experience, or in my situation many download errors are caused by insufficient storage space. 

Don't you use "download to external storage"? Or is this also already full on your phone? :D 


> I think this could be handled better if the storage space for a download would be checked before starting the download 

I guess this is not possible since zapp is using the android download manager and I can imagine zapp itself doesn't know the exact space of a video file before starting. Isn't it kind of "fast streaming" what the mediathek apps do? 

But I guess all of your question will be better answered be cemrich itself :)


> Please take all of this as **constructive** critics.

+1 from my side. I really like your app and I'm so glad that it exists. Keep up the good work

**Comment by justf0rfun** [comment]
> > In my experience, or in my situation many download errors are caused by insufficient storage space.
> 
> Don't you use "download to external storage"? Or is this also already full on your phone? :D

Yes, it is almost full. I have to do manual manage it like downloading a video and deleting it after watching. Also I the device does not have a SDCard expansion slot. So all storage is integral part of the device. I think the division between "internal" and "external" storage is just logical, maybe trough different partitions or so. I am not an android specialist.

> > I think this could be handled better if the storage space for a download would be checked before starting the download
> 
> I guess this is not possible since zapp is using the android download manager and I can imagine zapp itself doesn't know the exact space of a video file before starting. Isn't it kind of "fast streaming" what the mediathek apps do?

I think android does some download magic for optimization of battery and modem usage. I am not sure wether an app can see that. In general when downloading a video there is a file which gets completed chunk after chunk (i.e. by appending). After that you have a complete file on storage. When streaming a video I would expect that you never have a complete copy of the video (except very small ones) on your device, because every chunk will be downloaded for temporary use only and gets deleted after that. More sophisticated video players might hold some parts of the video for faster rewind. There is another way to download a file called sparse file where the full space of the complete file is allocated beforehand and then gets filled chunk after chunk in any order. This requires random file access.
 
When I look at mediathekviewweb.de there we can see the size of the download before we start it. I guess that the file size can be obtained by HTTP header from the server or so.

I never wrote an android app, so I am no expert and I had just a quick look at zapp's code. There is MediaStore in usage which doesn't seem to offer pre-allocation of files. 

Other findings independend from zapp:

The download manager API I found does not offer storage allocation.
https://developer.android.com/reference/android/app/DownloadManager

Discussion about sparse file implementation on android. Unclear if it works. It doesn't seem to use android specific APIs
https://stackoverflow.com/questions/9252332/android-pre-allocate-space-for-a-file-before-downloading-it#

There is an android StorageManager API which seems to offer pre-allocation for storage objects.
https://developer.android.com/reference/android/os/storage/StorageManager

I don't know if this is of any help. Also it might mean to rebuilt significant parts of zapp for pre-download-allocations.

**Comment by cemrich** [comment]
> Another question. Is it possible to change the download button back to the old behaviour where you could directly select the download quality?

Not right now. It clutters the UI too much for most users. Maybe I will introduce a "Remember" checkbox in the future but I shy away from making the UI more complex than it really has to be. Zapp should be easy to use and navigate for most users.

> Please take all of this as constructive critics.

I love hearing feedback from actual users :)

> Regarding the question, where to find the videos on the external storage, in my case the path is the following (should be similar on other devices)
> /storage/<name_of_external_sd_card_e.g._4707-08115>/Android/media/de.christinecoenen.code.zapp

This path will be different on Android 10 upwards. This is shy Android file handling is such a pain.,,

> I thought maybe there should be a "Downloads" view inside zapp.

This is planned - at least a tab with some kind of show "history" (started playing, downloaded, starred, whatever...). This is however no small feature and will require some time.

>  guess it is just more easy to use the already available download manager

Sadly, it is not. Google is busy with porting Android away from a file based system to some scoped storage stuff and forgot to move the build in DownloadManager along. As a result it cannot handle `content://` urls and therefore not download to sd card.  I am not very keen to implement download handling either, so I am using [fetch download library](https://github.com/tonyofrancis/Fetch) as an abstraction.

> I think this could be handled better if the storage space for a download would be checked before starting the download

The fetch download library is providing error messages when storage is full. I had to disable the file allocation before download because it caused downloads to the sd card to freeze quite a long time before they actually started, causing all kinds of problems.

I've implemented file downloads and -handling for other apps and sadly it is always a huge pain with a lot of trade offs.


@MasothU 
@justf0rfun 
Can you tell me what your use case is for video downloads? 
Do you use it exclusively for offline watching? 
Or do you pull the files from your device or use it otherwise?

I am thinking about using ExoPlayer (the video player zapp uses developed by Google) [for downloads](https://medium.com/google-exoplayer/downloading-streams-6d259eec7f95). 
Upside: All stream types (including ORF videos) can be downloaded. It is well maintained and usually rock solid. 
Downside: Videos cannot be accessed by any other app or the user because they are not saved as regular files.

**Comment by justf0rfun** [comment]
> > I thought maybe there should be a "Downloads" view inside zapp.
> 
> This is planned - at least a tab with some kind of show "history" (started playing, downloaded, starred, whatever...). This is however no small feature and will require some time.

That sounds really good. This way I can see what I have watched and not watched. 

> > Regarding the question, where to find the videos on the external storage, in my case the path is the following (should be similar on other devices)
> > /storage/<name_of_external_sd_card_e.g._4707-08115>/Android/media/de.christinecoenen.code.zapp
> 
> This path will be different on Android 10 upwards. This is shy Android file handling is such a pain.,,
> 
> > guess it is just more easy to use the already available download manager
> 
> Sadly, it is not. Google is busy with porting Android away from a file based system to some scoped storage stuff and forgot to move the build in DownloadManager along. As a result it cannot handle `content://` urls and therefore not download to sd card. I am not very keen to implement download handling either, so I am using [fetch download library](https://github.com/tonyofrancis/Fetch) as an abstraction.
> 
> > I think this could be handled better if the storage space for a download would be checked before starting the download
> 
> The fetch download library is providing error messages when storage is full. I had to disable the file allocation before download because it caused downloads to the sd card to freeze quite a long time before they actually started, causing all kinds of problems.
> 
> I've implemented file downloads and -handling for other apps and sadly it is always a huge pain with a lot of trade offs.

This paragraph will go semi off topic to map the big picture:
That sounds really scary. I mean the direction in which android goes. Even today there are so many people don't really understand files and file management. This is bad because it is a fundamental part of computing. Trying to hide that does not only mean that things get easier for people but also that people have less control and less possibilities. I don't think that is good. This sounds more like apple. Anyway there are Linux smartphones upcoming. I already own a pinephone since a couple of days and I and hopefully I will get a Librem 5 soon, too. Of course this is a niche market and if there is a chance that those devices gets success in bigger scale it will not be before the quality got better. Still android and ios devices are years ahead for some criteria while those Linux devices are ahead in terms of freedom, privacy, trust and control. Building Zapp for Linux with flexible UI for different screen sizes, would be a new topic.

> @MasothU
> @justf0rfun
> Can you tell me what your use case is for video downloads?
> Do you use it exclusively for offline watching?
> Or do you pull the files from your device or use it otherwise?

Beside live streaming I use zapp for downloading videos for these purposes:
1. For watching videos seamless while there is insufficient bandwidth or mobile connectivity. This could be at some places at home or when I go on a trip.
2. For archiving videos. Even if there are videos available for years in the mediathek there is no guarantee how long they will be there. Some videos are only available about a week. Some are really special or rare and have a certain value for me. So I do archive selected videos on my file server for personal use only. There was a phase were I did this very much but it takes to much time and to much storage space so I had to reduce my selection.

For point two it is essential to have file access or at least an export function. File access would of course be much easier. It offers endless possibilities for further usage or processing like take a video software and make a montage or so.

> I am thinking about using ExoPlayer (the video player zapp uses developed by Google) [for downloads](https://medium.com/google-exoplayer/downloading-streams-6d259eec7f95).
> Upside: All stream types (including ORF videos) can be downloaded. It is well maintained and usually rock solid.
> Downside: Videos cannot be accessed by any other app or the user because they are not saved as regular files.

Are ORF videos in m3u8 or so? In rare cases I did do downloads of videos from web players without download function and maybe from the mediathek to (don't remember exactly). Some video formats are just split up into chunks or slices which can be appended or concatenated in sequential order, I did that with cat on Linux. This results in a *.ts (Video Transport Stream File) file.

[German]
https://www.reviversoft.com/file-extensions/ts

**Comment by cemrich** [comment]
> That sounds really scary. 
Yes and no. It is a developer nightmare and layers upon layers of abstraction. But on the other hand Android is getting more secure with this sandbox approach.

> For point two it is essential to have file access or at least an export function.
There seems to be no way to get the file on a file system base in Android 10. However, you should be able to find and copy/delete/move/share it using any file manager app that is able to display collections of videos, music and images.
Would a "share file" option help, to improve interop with other apps?

> Are ORF videos in m3u8 or so?
Yes, but I will not implement this by hand :-/

**Comment by MasothU** [comment]
> Can you tell me what your use case is for video downloads?
> Do you use it exclusively for offline watching?
> Or do you pull the files from your device or use it otherwise?
> 
> I am thinking about using ExoPlayer (the video player zapp uses developed by Google) [for downloads](https://medium.com/google-exoplayer/downloading-streams-6d259eec7f95).
> Upside: All stream types (including ORF videos) can be downloaded. It is well maintained and usually rock solid.
> Downside: Videos cannot be accessed by any other app or the user because they are not saved as regular files.

Please, leave the download mechanism as it is! :) Here is why and my use cases with zapp/mediathek:
- I download a lot of episodes using zapp (at least 20 per week) and can't watch them all on my smartphone
- I sync them to OneDrive to stream them directly via BubbleUPnP to any TV with Chromecast/FireTV Stick in my house/at a friend and to free up space on my smartphone --> need direct access to the video files
- I sync them to my PC at home to watch episodes via Kodi on my TV (nature documentaries, big stuff) --> need direct access
- When I watch episodes on my smartphone I use exclusively another Videoplayer (MX Player). I like that Player, a lot of features (too many to list them :) for example faster playback speed. I watch some regurarly reportages with 130% speed. People tend to speak very slowly and still you get everything ;) --> need direct access to the Video files

As you can see, all of my use cases need direct access to the "real" Video files. I beg you, to not change this :)
I know, some of my use cases seem a little bit inconvient to you (download via zapp to sync to PC instead of using mediathek on  PC)
But this is very reasonable. I only use the PC/OneDrive to be able to watch them via Kodi on my big TV. And I'm not often at my PC directly. But with my Smartphone I can download new episodes right away from my couch :)

Sorry for this long "pamphlet" but when I found a solution to download mediathek right away with my smartphone I was so happy.

Being not able do access the videos directly, zapp would have not such a big advantage to me any more over the normal mediathek apps, since there you can also already download videos to watch them later.

**Comment by cemrich** [comment]
@MasothU Thanks for your feedback!

Zapp 3.6.0 is out which fixed another nasty download bug. Does this issue still persist?

**Comment by MasothU** [comment]
Today was my weekly download session. Except from some failed videos (perhaps due to Wifi disconnects) the error did not occur to me anymore.
Therefore, you can close the issue from my side (if no other user reports that this error still exists)
One last wish before we leave :)
An overview of the current running downloads, downloaded episodes and most important, the failed ones would be great.
I know failed ones appear in the notification area but sometimes you miss them or they vanish.

Thanks again for this great app and keep up the good work.

**Comment by cemrich** [comment]
Nice!

> An overview of the current running downloads, downloaded episodes and most important, the failed ones would be great.

This feature is planned but might take quite a while.
